cmd/kubeadm/app/cmd/join.go

```diff
@@ -619,6 +619,16 @@ func fetchInitConfigurationFromJoinConfiguration(cfg *kubeadmapi.JoinConfigurati
 		return nil, nil, err
 	}
 
+	// Create the final KubeConfig file with the cluster name discovered after fetching the cluster configuration
+	clusterinfo := kubeconfigutil.GetClusterFromKubeConfig(tlsBootstrapCfg)
+	tlsBootstrapCfg = kubeconfigutil.CreateWithToken(
```

Comment:
This doesn't work, as the client cert returned from TLS bootstrap would be overridden with the tls bootstrap token.
The way to do this is:
1. Extract `clusterinfo` (like you do)
2. Delete the existing cluster key and value in the .Clusters map
3. Replace the current context's `.Cluster` value
4. Add `clusterinfo` back to the `.Clusters` map with the new `initConfiguration.ClusterName` key

Reply:
@luxas Out of curiosity, in the case you are mentioning, this is where `discovery.For` is supposed to return? (so `len(cfg.Discovery.TLSBootstrapToken) == 0`): https://github.com/kubernetes/kubernetes/blob/91d6d7530382bf120a1575bae72ca36a33842e95/cmd/kubeadm/app/discovery/discovery.go#L45-L47